Add MCP connector

elvankonukseven opened the pull request:
This PR transforms the LLM4S framework from supporting only local tools to being compatible with the Model Context Protocol ecosystem, while maintaining perfect backward compatibility:

-MCPToolRegistry: Extends existing ToolRegistry for seamless Agent integration
-MCPClient: Handles JSON-RPC communication with MCP servers
-Transport Layer: Supports both stdio (subprocess) and SSE (HTTP) transports
-Schema Conversion: Automatically converts MCP tool schemas to framework format
- Tool Prioritization: Local tools tried first, MCP tools as fallback
- Existing Agent code works unchanged

The functionality has been tested through the MCPWeatherExample and MCPAgentExample files with a mock server: MCPSseServer. These examples also show how to start and connect to an MCP server.  

closes #27

elvankonukseven commented on samples/src/main/scala/org/llm4s/samples/mcp/MCPSseServer.scala:
```diff
@@ -0,0 +1,255 @@
+          val to = arguments("to").str.toUpperCase
+          
+          // Mock exchange rates
+          val rate = (from, to) match {
```
@rorygraves I have added the exception "Unsupported currency conversion from... to..." and the LLM's response is as follows on unknown currencies:

<img width="786" alt="Screenshot 2025-06-07 at 14 54 09" src="https://github.com/user-attachments/assets/b8c91b61-bae9-41d5-a3ac-197b98d3ac72" />

Would this be okay ?

rorygraves replied:
Ah, thats not really what I was thinking off.    This may be bad handling in the tool calling code I'm really trying to signal a tool side failure back to the llm with a useful error message.  I.e. the exception from the tool call should be returned in some kind of error wrapper so the LLM knows the method failed with an error - this makes it look like the system failed in some nasty way.

elvankonukseven replied:
@rorygraves I understand, I have tried an another approach and got :

<img width="854" alt="Screenshot 2025-06-07 at 17 03 54" src="https://github.com/user-attachments/assets/3d06862d-2526-4c0d-b7b3-f1d548ddfa8b" />

Is this what you had in mind ?

rorygraves replied:
Yes, thats much more what I would expect -  i think that we likely do not differentiate between tool exceptions and internal handling exceptions well enough - it should be clear i the repose to the LLM whether it is:
1) the method call throwing an error
2. the parameters being invalid
3. an internal LLM4S issue (e.g. container has died).

elvankonukseven commented on src/main/scala/org/llm4s/mcp/MCPToolRegistry.scala:
```diff
@@ -0,0 +1,129 @@
+class MCPToolRegistry(
+  mcpServers: Seq[MCPServerConfig],
+  localTools: Seq[ToolFunction[_, _]] = Seq.empty,
+  cacheTTL: Duration = 10.minutes 
```
Hi @rorygraves , I might be misunderstanding your suggestion but I think the cacheTTL is already a constructor param with default value ?

elvankonukseven commented:
Thanks a lot for all the suggestions @rorygraves ! I have implemented your suggestions, added overall more logging and moves the MCPAgentExample to the agent directory (in samples). 

Please let me know if there is anything else I could improve!
